cirq-rigetti/cirq_rigetti/aspen_device.py

```diff
@@ -0,0 +1,514 @@
+            if not (qubit.x <= self._number_octagons * 8):
+                raise UnsupportedQubit(
+                    'this Aspen device only supports line ',
+                    f'qubits up to length {self._number_octagons * 8}',
+                )
+            return
+
+        if isinstance(qubit, cirq.NamedQubit):
+            try:
+                index = int(qubit.name)
+                if not (index < self._maximum_qubit_number):
+                    raise UnsupportedQubit(
+                        'this Aspen device only supports qubits up to index '
+                        f'{self._maximum_qubit_number}'
+                    )
+                if not ((index % 10) <= 7):
+                    raise UnsupportedQubit(
+                        f'this Aspen device only supports qubit indices mod 10 <= 7'
+                    )
+                return
+
+            except ValueError:
+                raise UnsupportedQubit(
+                    f'Aspen devices only support named qubits by octagonal index'
+                )
+
+        if isinstance(qubit, (OctagonalQubit, AspenQubit)):
+            if not (qubit.index < self._maximum_qubit_number):
+                raise UnsupportedQubit(
+                    'this Aspen device only supports ',
+                    f'qubits up to index {self._maximum_qubit_number}',
+                )
+            return
+
+        else:  # pragma: no cover
+            raise UnsupportedQubit(f'unsupported Qid type {type(qubit)}')
+
+    def validate_operation(self, operation: 'cirq.Operation') -> None:
+        """Raises an exception if an operation does not satisfy the topological constraints
+        of the device.
+
+        Note, in case the operation is invalid, you can still use the Quil
+        compiler to rewire qubits and decompose the operation to this device's
+        topology.
+
+        Additionally, this method will not attempt to decompose the operation into this
+        device's native gate set. This integration, by default, uses the Quil
+        compiler to do so.
+
+        Please see the Quil Compiler
+        [documentation](https://pyquil-docs.rigetti.com/en/stable/compiler.html)
+        for more information.
+
+        Args:
+            operation: The operation to validate.
+
+        Raises:
+            UnsupportedRigettiQCSOperation: The operation isn't valid for this device.
+        """
+        qubits = operation.qubits
+        for qubit in qubits:
+            self.validate_qubit(qubit)
+        if len(qubits) == 2:
+            i = self._aspen_qubit_index(qubits[0])
+            j = self._aspen_qubit_index(qubits[1])
+            if j not in self.qubit_topology[i]:
+                raise UnsupportedRigettiQCSOperation(
+                    f'qubits {qubits[0]} and {qubits[1]} do not share an edge',
+                )
+
+    def _value_equality_values_(self):
+        return self._maximum_qubit_number
+
+    def __repr__(self):  # pragma: no cover
```

Comment:
use `# coverage: ignore`. Also probably not needed, if yes, I would prefer testing it.